### PR TITLE
Feat/signed attrs

### PR DESCRIPTION
**Pull request by vedhavyas**

### CODE REVIEW

File: documents/validator.go [hunk 1]
```diff
@@ -400,6 +400,33 @@ func anchorRepoAddressValidator(anchoredRepoAddr common.Address) Validator {
...
+// TODO(ved): tests
+func attributeValidator(idSrv identity.Service) Validator {
+	return ValidatorFunc(func(_, model Model) (err error) {
+		attrs := model.GetAttributes()
+		for _, attr := range attrs {
+			if attr.Value.Type != AttrSigned {
+				continue
+			}
+
+			signed := attr.Value.Signed
```
Comment (lucasvo) [inline documents/validator.go]: I would reuse the signature payload generation code from above here.

File: documents/validator.go [hunk 2]
```diff
@@ -400,6 +400,33 @@ func anchorRepoAddressValidator(anchoredRepoAddr common.Address) Validator {
 	})
 }
 
+// attributeValidator validates the signed attributes.
+// TODO(ved): tests
+func attributeValidator(idSrv identity.Service) Validator {
```
Comment (lucasvo) [inline documents/validator.go]: I assume you will still add this to the validator groups?

File: documents/attribute.go [hunk 3]
```diff
@@ -164,20 +188,56 @@ type Attribute struct {
...
+// takes keyLabel, signer identity, signer account, model and value
+// signature payload: sign(identity + docID + docVersion + value + time)
+func NewSignedAttribute(keyLabel string, identity identity.DID, account config.Account, model Model, value []byte) (attr Attribute, err error) {
 	attrKey, err := AttrKeyFromLabel(keyLabel)
 	if err != nil {
 		return attr, err
 	}
 
-	attrVal, err := AttrValFromString(attrType, value)
+	// TODO(ved): time should be part of the signature
```
Comment (lucasvo) [inline documents/attribute.go]: I don't think time needs to be added.

File: documents/attribute.go [hunk 4]
```diff
@@ -164,20 +188,56 @@ type Attribute struct {
...
+	var signPayload []byte
```
Comment (lucasvo) [inline documents/attribute.go]: Can you create a helper method that takes a Signed field struct and constructs the payload from it and then reuse it below?

File: documents/attribute.go (hunk 3 above)
Comment (vedhavyas) [reply documents/attribute.go]: yeah. I just realised it

File: documents/validator.go (hunk 2 above)
Comment (vedhavyas) [reply documents/validator.go]: not complete yet

File: documents/attribute.go (hunk 3 above)
Comment (vedhavyas) [reply documents/attribute.go]: I still want to think on this more.

File: documents/validator.go [hunk 5]
```diff
@@ -400,6 +400,44 @@ func anchorRepoAddressValidator(anchoredRepoAddr common.Address) Validator {
...
 
+// attributeValidator validates the signed attributes.
+func attributeValidator(repo anchors.AnchorRepository, idSrv identity.Service) Validator {
+	return ValidatorFunc(func(_, model Model) (err error) {
+		ts, err := model.Timestamp()
+		if err != nil {
+			return err
+		}
+
+		attrs := model.GetAttributes()
```
Comment (mikiquantum) [inline documents/validator.go]: Would it be interesting to have a getter in the model for all SignedAttributes?

File: documents/validator.go [hunk 6]
```diff
@@ -400,6 +400,44 @@ func anchorRepoAddressValidator(anchoredRepoAddr common.Address) Validator {
...
+			signed := attr.Value.Signed
+			var aid anchors.AnchorID
+			aid, err = anchors.ToAnchorID(signed.DocumentVersion)
+			if err != nil {
+				return err
+			}
+
+			_, ats, erro := repo.GetAnchorData(aid)
+			if erro != nil {
+				// maybe not anchored yet
```
Comment (mikiquantum) [inline documents/validator.go]:
```suggestion
				// maybe not anchored yet as we validate signatures a several times before anchoring
```

File: documents/attribute.go [hunk 7]
```diff
@@ -96,14 +118,11 @@ type AttrVal struct {
 	Str       string
 	Bytes     []byte
 	Timestamp *timestamp.Timestamp
+	Signed    Signed
```
Comment (mikiquantum) [inline documents/attribute.go]: Having a new type, implies that we only support []byte values and we will not need to perform any specific/typed operation over values of type `signed`?

Comment (vedhavyas) [reply documents/attribute.go]: not sure I understand this

File: documents/validator.go (hunk 5 above)
Comment (vedhavyas) [reply documents/validator.go]: The idea to have a single function to give all attributes and keep the model API simple

**Pull request merged**